src/plugins/canonical/canonical.c

```diff
@@ -0,0 +1,337 @@
+/**
+ * @file
+ *
+ * @brief Source for canonical plugin
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ *
+ */
+
+#include "canonical.h"
+
+#include <fnmatch.h>
+#include <kdbease.h> //elektraReadArrayNumber
+#include <kdbhelper.h>
+#include <kdbmeta.h>    //elektraMetaArrayToKS
+#include <kdbos.h>      //ELEKTRA_MAX_ARRAY_SIZE
+#include <kdbprivate.h> //elektraArrayValidateName
+#include <kdbtypes.h>   //elektra_long_long_t
+#include <kdbutility.h> //elektraLskip, elektraRstrip
+#include <regex.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h> //strcasecmp
+
+
+typedef enum {
+	REGEX = (1 << 0),
+	FNMATCH = (1 << 1),
+	LISTSENSITIVE = (1 << 2),
+	LISTINSENSITIVE = (1 << 3),
+	WORD = (1 << 4),
+} TransformType;
+
+static char ** stringToArray (const char * string, const char * delim)
+{
+	if (!string || !delim) return NULL;
+
+	char ** array = NULL;
+	int elems = 1;
+	const char * ptr = string;
+
+	while (*ptr)
+	{
+		if (*ptr == *delim)
+		{
+			++elems;
+		}
+		++ptr;
+	}
+	++elems; // terminating element
+	array = elektraCalloc (elems * sizeof (char *));
+	char * localString = NULL;
+	if (string[0] == '[' || string[0] == '(' || string[0] == '{')
+		localString = elektraStrDup (string + 1);
```

Comment:
minor: you could avoid the else branch by calculating an offset (which is either 0 or 1)

Please prefer to use `{}`.